File: src/main/mods/data/domain/web-token.ts
import type { DataWebTokenSnapshot, WebTokenClaims } from './types'

/**
 * 从 webToken 中提取 xbox-webapi 所需声明
 */
export function resolveWebTokenClaims(raw: DataWebTokenSnapshot): WebTokenClaims | null {
  const token = raw
  const userToken = token.data?.Token ?? token.Token
  const uhs = token.data?.DisplayClaims?.xui?.[0]?.uhs ?? token.DisplayClaims?.xui?.[0]?.uhs
  if (typeof userToken !== 'string' || userToken.length === 0) {
    return null
  }
  if (typeof uhs !== 'string' || uhs.length === 0) {
    return null
  }

  return {
    userToken,
    uhs
  }
}

File: tauri/src-tauri/src/mods/data/domain/web_token.rs
use serde_json::Value;

#[derive(Debug, Clone)]
pub struct WebTokenClaims {
    pub user_token: String,
    pub uhs: String,
}

// 从 webToken 快照中提取 xbox-webapi 初始化所需声明。
pub fn resolve_web_token_claims(raw: &Value) -> Option<WebTokenClaims> {
    let token = raw.get("data").unwrap_or(raw);

    let user_token = token.get("Token").and_then(|value| value.as_str())?;
    let uhs = token
        .get("DisplayClaims")
        .and_then(|value| value.get("xui"))
        .and_then(|value| value.as_array())
        .and_then(|xui| xui.first())
        .and_then(|value| value.get("uhs"))
        .and_then(|value| value.as_str())?;

    if user_token.is_empty() || uhs.is_empty() {
        return None;
    }

    Some(WebTokenClaims {
        user_token: user_token.to_string(),
        uhs: uhs.to_string(),
    })
}
